src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -517,66 +509,30 @@ private ScoreOrderedResultRetriever(CloseableIterator<ScoredPrimaryKey> scoredPr
         {
             @SuppressWarnings("resource")
-            UnfilteredRowIterator iterator = nextRowIterator(this::nextSelectedKeyInRange);
+            UnfilteredRowIterator iterator = nextRowIterator();
             // Because we know ordered keys are fully qualified, we do not iterate partitions
             return iterator != null ? iterator : endOfData();
         }
 
         /**
-         * Tries to obtain a row iterator for one of the supplied keys by repeatedly calling
-         * {@link ResultRetriever#apply} until it gives a non-null result.
-         * The keySupplier should return the next key with every call to get() and
-         * null when there are no more keys to try.
+         * Tries to obtain a row iterator for the next valid row or returns null if no more rows are available.
          *
          * @return an iterator or null if all keys were tried with no success
          */
-        private @Nullable UnfilteredRowIterator nextRowIterator(@Nonnull Supplier<ScoredPrimaryKey> keySupplier)
-        {
-            UnfilteredRowIterator iterator = null;
-            while (iterator == null)
-            {
-                ScoredPrimaryKey key = keySupplier.get();
-                if (key == null)
-                    return null;
-                iterator = apply(key);
-            }
-            return iterator;
-        }
-
-        /**
-         * Determine if the key is in one of the queried key ranges. We do not iterate through results in
-         * {@link PrimaryKey} order, so we have to check each range.
-         * @param key
-         * @return true if the key is in one of the queried key ranges
-         */
-        private boolean isInRange(DecoratedKey key)
-        {
-            if (coversFullRing)
-                return true;
-
-            for (AbstractBounds<PartitionPosition> range : keyRanges)
-                if (range.contains(key))
-                    return true;
-            return false;
-        }
-
-        /**
-         * Returns the next available key contained by one of the keyRanges and selected by the queryController.
-         * If the next key falls out of the current key range, it skips to the next key range, and so on.
-         * If no more keys acceptd by the controller are available, returns null.
-         */
-        private @Nullable ScoredPrimaryKey nextSelectedKeyInRange()
+        private @Nullable UnfilteredRowIterator nextRowIterator()
         {
             while (scoredPrimaryKeyIterator.hasNext())
             {
-                ScoredPrimaryKey key = scoredPrimaryKeyIterator.next();
-                if (isInRange(key.partitionKey()) && controller.selects(key))
-                    return key;
+                // We do not check if the PrimaryKey is in the DataRanges for the query because materialization from
+                // storage does that for us.
+                UnfilteredRowIterator iterator = materializeRowAndApplyFilter(scoredPrimaryKeyIterator.next());
```

Comment:
This reasoning is incomplete based on the current SAI design because we don't push all token ranges down into the vector indexes.